src/node_messaging.cc

```diff
@@ -128,12 +175,62 @@ class SerializerDelegate : public ValueSerializer::Delegate {
     env_->isolate()->ThrowException(Exception::Error(message));
   }
 
+  Maybe<bool> WriteHostObject(Isolate* isolate, Local<Object> object) override {
+    if (env_->message_port_constructor_template()->HasInstance(object)) {
+      return WriteMessagePort(Unwrap<MessagePort>(object));
+    }
+
+    env_->ThrowError("Cannot serialize unknown type of host object");
+    return Nothing<bool>();
+  }
+
+  Maybe<uint32_t> GetSharedArrayBufferId(
+      Isolate* isolate, Local<SharedArrayBuffer> shared_array_buffer) override {
+    uint32_t i;
+    for (i = 0; i < seen_shared_array_buffers_.size(); ++i) {
+      if (seen_shared_array_buffers_[i] == shared_array_buffer)
+        return Just(i);
+    }
+
+    SharedArrayBufferMetadataReference reference(
+        SharedArrayBufferMetadata::ForIncomingSharedArrayBuffer(env_,
+                                                          context_,
+                                                          shared_array_buffer));
+    if (!reference) {
+      return Nothing<uint32_t>();
+    }
+    seen_shared_array_buffers_.push_back(shared_array_buffer);
+    msg_->AddSharedArrayBuffer(reference);
+    return Just(i);
+  }
+
+  void Finish() {
+    for (MessagePort* port : ports_) {
+      port->Close();
+      msg_->AddMessagePort(port->Detach());
+    }
+  }
+
   ValueSerializer* serializer = nullptr;
 
  private:
+  Maybe<bool> WriteMessagePort(MessagePort* port) {
+    for (uint32_t i = 0; i < ports_.size(); i++) {
+      if (ports_[i] == port) {
+        serializer->WriteUint32(i);
+        return Just(true);
+      }
+    }
+
+    env_->ThrowError("MessagePort was not listed in transferList");
```

Comment:
Test needed. Also TypeError.

Reply:
Hm, at least Firefox throws a special `DataCloneError`, which is a `DOMException` and therefore a plain `Error`…

Reply:
DOMException isn't really encouraged for new errors any more. TypeErrors and other error types that map to JS are the future for the web. Plus we don't really care about compatibility, do we?

Reply:
Yeah, not really the point :) Still, why `TypeError`? I don’t really think the semantics here are “this thing has the wrong type”, more like, “you missed an entry in a list, please add it”, which doesn’t seem like a type mismatch to me…

> Plus we don't really care about compatibility, do we?

Well … for `MessagePort`s/`MessageChannel`s we kind of do? :smile: Maybe not in this first iteration but it should be doable to get some degree of compatibility

Reply:
> Still, why `TypeError`?

> TypeError is used to indicate an unsuccessful operation when none of the other *NativeError* objects are an appropriate indication of the failure cause.
>
> https://tc39.github.io/ecma262/#sec-native-error-types-used-in-this-standard-typeerror

Either way I just realized `require('v8').Serializer.prototype._getDataCloneError` is also `Error`, so I guess consistency trumps all.

Reply:
> > https://tc39.github.io/ecma262/#sec-native-error-types-used-in-this-standard-typeerror

I think that’s because the language spec isn’t really concerned with these kinds of errors?

> Either way I just realized `require('v8').Serializer.prototype._getDataCloneError` is also `Error`, so I guess consistency trumps all.

I mean … that’s because I wrote it that way. :smile: We definitely *can* change that if you think it makes sense